lib/go-atscfg/headerrewritedotconfig.go

```diff
@@ -23,74 +23,300 @@ import (
 	"errors"
 	"fmt"
 	"math"
+	"net/url"
 	"regexp"
 	"strconv"
 	"strings"
 
 	"github.com/apache/trafficcontrol/lib/go-tc"
-	"github.com/apache/trafficcontrol/lib/go-util"
 )
 
 const HeaderRewritePrefix = "hdr_rw_"
+const HeaderRewriteMidPrefix = "hdr_rw_mid_"
 const ContentTypeHeaderRewriteDotConfig = ContentTypeTextASCII
 const LineCommentHeaderRewriteDotConfig = LineCommentHash
 
 const ServiceCategoryHeader = "CDN-SVC"
 
 const MaxOriginConnectionsNoMax = 0 // 0 indicates no limit on origin connections
 
+const HeaderRewriteFirstPrefix = HeaderRewritePrefix + "first_"
+const HeaderRewriteInnerPrefix = HeaderRewritePrefix + "inner_"
+const HeaderRewriteLastPrefix = HeaderRewritePrefix + "last_"
+
+func FirstHeaderRewriteConfigFileName(dsName string) string {
+	return HeaderRewriteFirstPrefix + dsName + ConfigSuffix
+}
+
+func InnerHeaderRewriteConfigFileName(dsName string) string {
+	return HeaderRewriteInnerPrefix + dsName + ConfigSuffix
+}
+
+func LastHeaderRewriteConfigFileName(dsName string) string {
+	return HeaderRewriteLastPrefix + dsName + ConfigSuffix
+}
+
+// MakeHeaderRewriteDotConfig makes the header rewrite file for
+// an Edge hdr_rw_ or Mid hdr_rw_mid_ or Topology hdr_rw_{first,inner,last} file,
+// as generated by MakeMetaConfigFilesList.
 func MakeHeaderRewriteDotConfig(
 	fileName string,
 	deliveryServices []DeliveryService,
 	deliveryServiceServers []tc.DeliveryServiceServer,
 	server *Server,
 	servers []Server,
+	cacheGroupsArr []tc.CacheGroupNullable,
+	tcServerParams []tc.Parameter,
+	serverCapabilities map[int]map[ServerCapability]struct{},
+	requiredCapabilities map[int]map[ServerCapability]struct{},
+	topologiesArr []tc.Topology,
 	hdrComment string,
 ) (Cfg, error) {
 	warnings := []string{}
+	if server.CDNName == nil {
+		return Cfg{}, makeErr(warnings, "this server missing CDNName")
+	} else if tc.CacheType(server.Type) == tc.CacheTypeInvalid {
+		return Cfg{}, makeErr(warnings, "this server missing Type")
+	} else if server.HostName == nil {
+		return Cfg{}, makeErr(warnings, "server missing HostName")
+	} else if server.Cachegroup == nil {
+		return Cfg{}, makeErr(warnings, "server missing Cachegroup")
+	}
 
-	dsName := strings.TrimSuffix(strings.TrimPrefix(fileName, HeaderRewritePrefix), ConfigSuffix) // TODO verify prefix and suffix? Perl doesn't
+	cacheGroups, err := makeCGMap(cacheGroupsArr)
+	if err != nil {
+		return Cfg{}, makeErr(warnings, "making CacheGroup map: "+err.Error())
+	}
+	topologies := makeTopologyNameMap(topologiesArr)
+
+	// TODO verify prefix and suffix? Perl doesn't
+	dsName := fileName
+	dsName = strings.TrimSuffix(dsName, ConfigSuffix)
+	dsName = strings.TrimPrefix(dsName, HeaderRewriteFirstPrefix)
+	dsName = strings.TrimPrefix(dsName, HeaderRewriteInnerPrefix)
+	dsName = strings.TrimPrefix(dsName, HeaderRewriteLastPrefix)
+	dsName = strings.TrimPrefix(dsName, HeaderRewriteMidPrefix)
+	dsName = strings.TrimPrefix(dsName, HeaderRewritePrefix)
```

Comment:
This code means that someone who can create or modify a DS, possibly with scope limited by tenancy can alter the config for a different DS by manipulating the name of the DS. This feels like a potential security issue in the making.

Reply:
> Looks like underscores in the DS name will cause all sorts of other issues and just need to be removed entirely. This issue isn't unique to this, so we can merge.

Yep. I made https://github.com/apache/trafficcontrol/issues/5658 to add that validation to Traffic Ops